src/future/shared.rs

```diff
@@ -0,0 +1,229 @@
+            Err(error) => Err(error),
+        }
+    }
+
+    /// Clones the result from self.inner.state.
+    /// Assumes state is `State::Done`.
+    fn read_result(&self) -> Result<Async<SharedItem<F::Item>>, SharedError<F::Error>> {
+        match *self.inner.state.read().unwrap() {
+            State::Done(ref result) => Self::result_to_polled_result(result.clone()),
+            State::Waiting(_) => panic!("read_result() was called but State is not Done"),
+        }
+    }
+
+    /// Stores the result in self.inner.state, unparks the waiting tasks,
+    /// and returns the result.
+    fn store_result(&self,
+                    result: Result<SharedItem<F::Item>, SharedError<F::Error>>)
+                    -> Result<Async<SharedItem<F::Item>>, SharedError<F::Error>> {
+        let ref mut state = *self.inner.state.write().unwrap();
+
+        match mem::replace(state, State::Done(result.clone())) {
+            State::Waiting(waiters) => {
+                self.inner.result_ready.store(true, Ordering::Relaxed);
+                for task in waiters {
+                    task.unpark();
+                }
+            }
+            State::Done(_) => panic!("store_result() was called twice"),
+        }
+
+        Self::result_to_polled_result(result)
+    }
+}
+
+impl<F> Future for Shared<F>
+    where F: Future
+{
+    type Item = SharedItem<F::Item>;
+    type Error = SharedError<F::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // The logic is as follows:
+        // 1. Check if the result is ready (with result_ready)
+        //  - If the result is ready, return it.
+        //  - Otherwise:
+        // 2. Try lock the self.inner.original_future:
+        //    - If successfully locked, check again if the result is ready.
+        //      If it's ready, just return it.
+        //      Otherwise, poll the original future.
+        //      If the future is ready, unpark the waiting tasks from
+        //      self.inner.state and return the result.
+        //    - If the future is not ready, or if the lock failed:
+        // 3. Lock the state for write.
+        // 4. If the state is `State::Done`, return the result. Otherwise:
+        // 5. Create a task, push it to the waiters vector, and return `Ok(Async::NotReady)`.
+
+        // If the result is ready, just return it
+        if self.inner.result_ready.load(Ordering::Relaxed) {
+            return self.read_result();
+        }
+
+        // The result was not ready.
+        // Try lock the original future.
+        match self.inner.original_future.try_lock() {
+            Some(mut original_future) => {
+                // Other thread could already poll the result, so we check if result_ready.
+                if self.inner.result_ready.load(Ordering::Relaxed) {
+                    return self.read_result();
+                }
+
+                match original_future.poll() {
+                    Ok(Async::Ready(item)) => {
```

Comment:
It's generally good hygiene to run destructors associated with futures ASAP as it ensure that any remaining resources held internally are all deallocated and/or pending futures that we're not interested in are canceled.